src/inputfetcher.h

```diff
@@ -0,0 +1,174 @@
+ *
+ * It spawns a fixed set of worker threads that fetch Coins for each input
+ * in a block. The Coin is moved into the Input struct and then the ready flag
+ * is atomically updated to true. The main thread waits on the ready flag
+ * until it is true and then inserts it into the temporary cache.
+ *
+ * Worker threads are synchronized with the main thread using a barrier, which
+ * is used at the beginning of fetching to start the workers and at the end to
+ * make sure all workers have exited the work loop.
+ */
+class InputFetcher
+{
+private:
+    //! The latest input being fetched. Workers atomically increment this when fetching.
+    std::atomic_size_t m_input_head{0};
+
+    //! The inputs of the block which is being fetched.
+    struct Input {
+        //! Workers set this after setting the coin. The main thread tests this before reading the coin.
+        std::atomic_flag ready{};
+        //! The outpoint to fetch;
+        const COutPoint& outpoint;
+        //! The coin that workers will fetch and main thread will insert into cache.
+        Coin coin{};
+
+        Input(Input&& other) noexcept : outpoint{other.outpoint} {} // Only moved in setup for reallocation.
+        explicit Input(const COutPoint& o LIFETIMEBOUND) noexcept : outpoint{o} {}
+    };
+    std::vector<Input> m_inputs{};
+
+    /**
+     * The set of first 8 bytes of txids of all txs in the block being fetched.
+     * Used to filter out inputs that are created and spent in the same block,
+     * since they will not be in the db or the cache.
+     */
+    std::vector<uint64_t> m_txids{};
+
+    //! DB coins view to fetch from.
+    const CCoinsView* m_db{nullptr};
+    //! The cache to fetch from.
+    const CCoinsViewCache* m_cache{nullptr};
+
+    std::vector<std::thread> m_worker_threads{};
+    std::barrier<> m_barrier;
+    bool m_request_stop{false};
+
+    /**
+     * Fetches the next input in the queue. Safe to call from any thread once inside the barrier.
+     * 
+     * @return true if there are more inputs in the queue to fetch
+     * @return false if there are no more inputs in the queue to fetch
+     */
+    bool FetchCoin() noexcept
+    {
+        const size_t i{m_input_head.fetch_add(1, std::memory_order_relaxed)};
+        if (i >= m_inputs.size()) [[unlikely]] return false;
+        auto& input{m_inputs[i]};
+        if (std::binary_search(m_txids.begin(), m_txids.end(), input.outpoint.hash.ToUint256().GetUint64(0))) {
+            input.ready.test_and_set(std::memory_order_relaxed);
+            input.ready.notify_one();
+            return true;
+        }
+        auto coin{m_cache->GetPossiblySpentCoinFromCache(input.outpoint)};
+        if (!coin) {
+            try {
+                coin = m_db->GetCoin(input.outpoint);
+            } catch (const std::runtime_error& e) {
+                LogPrintLevel(BCLog::VALIDATION, BCLog::Level::Warning, "InputFetcher failed to fetch input: %s.\n", e.what());
```

Comment:
nit: trailing newline shouldn't be needed anymore

```suggestion
                LogPrintLevel(BCLog::VALIDATION, BCLog::Level::Warning, "InputFetcher failed to fetch input: %s.", e.what());
```